## 사전 데이터 분석

----------------
<span style="color:yellow">1. pandas 로 데이터 읽기<span>

In [2]:
import pandas as pd
df19 = pd.read_csv('./data/1. 19년도-과정이수및평가점수.csv', encoding='euc-kr')
df15 = pd.read_excel('./data/2. 15-19년전체집계.xlsx')

In [3]:
df19.head()

학생구분       학과  학년  반  A과목이수  B과목이수  C과목이수  D과목이수  E과목이수  A과목점수  B과목점수  \
0  신입생  건축설비소방과   1  1      2      2      2      2      2      2      5   
1  신입생  건축설비소방과   1  1      2      2      2      2      2      5      3   
2  신입생  건축설비소방과   1  1      2      1      2      2      2      5      1   
3  신입생  건축설비소방과   1  1      2      2      2      2      2      2      5   
4  신입생  건축설비소방과   1  1      1      1      1      1      1      2      5   

   C과목점수  D과목점수  E과목점수  
0      4      4      4  
1      3      7      3  
2      2      3      3  
3      5      3      4  
4      4      4      4

In [4]:
df15.head()

연도         학과 학생구분  응시자수       A과목       B과목  F과목       D과목  G과목  \
0  2019    건축설비소방과  신입생    70  3.028571  4.514286  0.0  3.400000  0.0   
1  2019       건축학부  신입생   229  2.982533  4.903930  0.0  3.489083  0.0   
2  2019       경영학부  신입생   147  2.829932  4.544218  0.0  3.993197  0.0   
3  2019  글로벌외식조리학부  신입생   147  3.136054  4.265306  0.0  3.476190  0.0   
4  2019        기계과  신입생    87  3.022989  4.321839  0.0  3.229885  0.0   

        C과목       E과목  
0  3.914286  3.728571  
1  3.572052  3.589520  
2  3.659864  3.741497  
3  3.700680  3.829932  
4  3.701149  3.747126

------------------------------------
<span style="color:yellow">2. 결측치 (missing value) 확인</span>

.shape : 데이터프레임의 행과 열 수를 알려준다  
.count() : 각 열 당 결측치를 제외한 행 수를 알려준다

In [5]:
print(df19.shape)
df19.count()

(3477, 14)


학생구분     3477
학과       3477
학년       3477
반        3477
A과목이수    3477
B과목이수    3477
C과목이수    3477
D과목이수    3477
E과목이수    3477
A과목점수    3477
B과목점수    3477
C과목점수    3477
D과목점수    3477
E과목점수    3477
dtype: int64

In [6]:
print(df15.shape)
df15.count()

(278, 11)


연도      278
학과      278
학생구분    278
응시자수    278
A과목     278
B과목     278
F과목     278
D과목     278
G과목     278
C과목     278
E과목     278
dtype: int64

결론 :  
* .shape 과 .count() 의 행 수가 일치하므로 적어도 표면적으로는 결측치가 없다.  
* df19 자료에서 일부 응시자의 점수가 전부 0인 것을 볼 때 자료 제공자의 전처리에 의해 결측치가 0으로 치환되었을 가능성이 있지만 주어진 자료만 가지고는 알 수 없다.  
* df19 자료에 F, G 과목이 없고 df15 자료의 2019 년도 부분에서 F, G 과목이 전부 0점 처리된 점, df15 자료의 2015~2018 년도 부분에서 F, G 과목 성적이 있지만 반대로 C, E 과목이 전부 0점 처리된 점으로 보아 2019년을 기점으로 응시과목이 F, G 에서 C, E 과목으로 변경되었을 가능성이 높다. 즉 df15 자료에서 단체로 0점 처리된 부분은 사실은 결측치이지만 자료 제공자의 전처리에 의해 결측치가 0으로 치환되어 제공되었을 가능성이 높다.

---------------------
<span style="color:yellow">3. df19 자료와 df15 자료의 2019 년도 부분 정합성 (응시자 수) 확인</span>

In [7]:
df19[df19['학생구분']=='신입생'].groupby(['학과']).학생구분.count()

학과
건축설비소방과       70
건축학부         229
경영학부         147
글로벌외식조리학부    147
기계과           87
디지털전자과        75
메카트로닉스과       66
방송음향영상학부     112
비서사무행정과       77
사회복지과         73
스포츠지도과        39
아동문헌정보과       62
언어재활과         42
유아교육과         72
의공융합과         31
자동차학부        111
자동화시스템과       63
전기과           76
전자통신과         75
컴퓨터정보학부      136
토목환경과         58
항공호텔관광학부     160
Name: 학생구분, dtype: int64

In [8]:
df15[((df15['학생구분']=='신입생') & (df15['연도']==2019))][['학과', '응시자수']]

학과  응시자수
0     건축설비소방과    70
1        건축학부   229
2        경영학부   147
3   글로벌외식조리학부   147
4         기계과    87
5      디지털전자과    75
6     메카트로닉스과    66
7    방송음향영상학부   112
8     비서사무행정과    77
9       사회복지과    73
10     스포츠지도과    39
11    아동문헌정보과    62
12      언어재활과    42
13      유아교육과    72
14      의공융합과    31
15      자동차학부   111
16    자동화시스템과    63
17        전기과    76
18      전자통신과    75
19    컴퓨터정보학부   136
20      토목환경과    58
21   항공호텔관광학부   160

df19 자료와 df15 자료의 2019 년도 부분 신입생 응시자 수가 같음을 확인.

In [43]:
df19[df19['학생구분']=='재학생'].groupby(['학과']).학생구분.count()

학과
건축과           63
건축설비소방과       59
경영과           67
기계과           55
디지털전자과        56
메카트로닉스과       28
모바일인터넷과       47
방송음향영상과       55
비서사무행정과       53
사회복지과         52
산업경영과         25
세무회계과         61
스포츠지도과        53
실내디자인과        60
아동문헌정보과       51
언어재활과         37
유아교육과         60
의공융합과         29
자동차과         121
자동화시스템과       36
전기과           48
전자통신과         41
컴퓨터소프트웨어과     58
토목환경과         25
항공서비스과        63
호텔관광과         54
호텔외식서비스과      40
호텔조리과         72
Name: 학생구분, dtype: int64

In [10]:
df15[((df15['학생구분']=='재학생') & (df15['연도']==2019))][['학과', '응시자수']]

학과  응시자수
22         건축과    63
23     건축설비소방과    59
24         경영과    67
25         기계과    55
26      디지털전자과    56
27     메카트로닉스과    28
28     모바일인터넷과    47
29     방송음향영상과    55
30     비서사무행정과    53
31       사회복지과    52
32       산업경영과    25
33       세무회계과    61
34      스포츠지도과    53
35  실내디자인과(학부)    60
36     아동문헌정보과    51
37       언어재활과    37
38       유아교육과    60
39       의공융합과    29
40        자동차과   121
41     자동화시스템과    36
42         전기과    48
43       전자통신과    41
44   컴퓨터소프트웨어과    58
45       토목환경과    25
46      항공서비스과    63
47       호텔관광과    54
48    호텔외식서비스과    40
49       호텔조리과    72

df19 자료와 df15 자료의 2019 년도 부분 재학생 응시자 수도 같음을 확인.  
(df19 자료엔 실내디자인과, df15 자료의 2019 년도 부분엔 실내디자인과(학부) 로 표기가 다름에 유의)

---------------------
<span style="color:yellow">4. df19 자료와 df15 자료의 2019 년도 부분 정합성 (성적) 확인</span>

In [165]:
score1_mean19 = df19[df19['학생구분']=='신입생'].groupby(['학과'])[df19.columns[-5:]].mean().round(6)
print(score1_mean19)

              A과목점수     B과목점수     C과목점수     D과목점수     E과목점수
학과                                                         
건축설비소방과    3.028571  4.514286  3.914286  3.400000  3.728571
건축학부       2.982533  4.903930  3.572052  3.489083  3.589520
경영학부       2.829932  4.544218  3.659864  3.993197  3.741497
글로벌외식조리학부  3.136054  4.265306  3.700680  3.476190  3.829932
기계과        3.022989  4.321839  3.701149  3.229885  3.747126
디지털전자과     2.773333  4.373333  3.626667  3.560000  3.720000
메카트로닉스과    2.954545  4.606061  3.863636  3.439394  3.636364
방송음향영상학부   3.428571  5.285714  3.955357  3.910714  4.053571
비서사무행정과    3.272727  5.207792  4.480519  3.870130  4.103896
사회복지과      2.945205  4.534247  3.520548  3.191781  3.561644
스포츠지도과     2.820513  4.307692  3.307692  3.102564  3.538462
아동문헌정보과    3.451613  4.983871  3.693548  4.064516  4.064516
언어재활과      3.023810  4.690476  4.238095  3.547619  3.880952
유아교육과      3.625000  5.222222  3.722222  3.875000  4.375000
의공융합과      2.870968  4.935484  4.000000 

In [166]:
score1_mean15 = df15[((df15['학생구분']=='신입생') & (df15['연도']==2019))].iloc[:,[1, 4, 5, 9, 7, 10]]
score1_mean15 = score1_mean15.set_index('학과').rename(columns=dict(zip(score1_mean15.columns[-5:], df19.columns[-5:])))
print(score1_mean15)

              A과목점수     B과목점수     C과목점수     D과목점수     E과목점수
학과                                                         
건축설비소방과    3.028571  4.514286  3.914286  3.400000  3.728571
건축학부       2.982533  4.903930  3.572052  3.489083  3.589520
경영학부       2.829932  4.544218  3.659864  3.993197  3.741497
글로벌외식조리학부  3.136054  4.265306  3.700680  3.476190  3.829932
기계과        3.022989  4.321839  3.701149  3.229885  3.747126
디지털전자과     2.773333  4.373333  3.626667  3.560000  3.720000
메카트로닉스과    2.954545  4.606061  3.863636  3.439394  3.636364
방송음향영상학부   3.428571  5.285714  3.955357  3.910714  4.053571
비서사무행정과    3.272727  5.207792  4.480519  3.870130  4.103896
사회복지과      2.945205  4.534247  3.520548  3.191781  3.561644
스포츠지도과     2.820513  4.307692  3.307692  3.102564  3.538462
아동문헌정보과    3.451613  4.983871  3.693548  4.064516  4.064516
언어재활과      3.023810  4.690476  4.238095  3.547619  3.880952
유아교육과      3.625000  5.222222  3.722222  3.875000  4.375000
의공융합과      2.870968  4.935484  4.000000 

In [167]:
score1_mean19.equals(score1_mean15)

True

df19 자료와 df15 자료의 2019 년도 부분 신입생 과목 평균점수가 같음을 확인.

In [161]:
score2_mean19 = df19[df19['학생구분']=='재학생'].groupby(['학과'])[df19.columns[-5:]].mean().round(6)
print(score2_mean19)

              A과목점수     B과목점수     C과목점수     D과목점수     E과목점수
학과                                                         
건축과        3.206349  4.301587  3.777778  3.603175  3.761905
건축설비소방과    3.152542  5.644068  4.016949  4.813559  4.389831
경영과        2.850746  4.880597  4.223881  4.432836  3.880597
기계과        3.109091  4.745455  3.836364  3.490909  2.727273
디지털전자과     2.857143  4.196429  3.875000  4.089286  4.071429
메카트로닉스과    2.607143  5.071429  3.857143  4.571429  4.107143
모바일인터넷과    3.042553  5.361702  4.361702  4.595745  4.085106
방송음향영상과    3.872727  5.327273  4.127273  4.200000  4.127273
비서사무행정과    3.226415  5.773585  4.792453  4.754717  4.452830
사회복지과      3.384615  5.076923  3.769231  4.288462  3.730769
산업경영과      3.080000  5.320000  4.720000  4.000000  4.120000
세무회계과      3.737705  5.032787  4.229508  4.459016  4.180328
스포츠지도과     3.056604  4.245283  3.660377  3.283019  3.396226
실내디자인과     3.416667  4.933333  4.383333  3.983333  4.400000
아동문헌정보과    3.725490  5.431373  4.019608 

In [162]:
score2_mean15 = df15[((df15['학생구분']=='재학생') & (df15['연도']==2019))].iloc[:,[1, 4, 5, 9, 7, 10]]
score2_mean15 = score2_mean15.set_index('학과').rename(columns=dict(zip(score2_mean15.columns[-5:], df19.columns[-5:])))
score2_mean15 = score2_mean15.rename(index={'실내디자인과(학부)':'실내디자인과'})
print(score2_mean15)

              A과목점수     B과목점수     C과목점수     D과목점수     E과목점수
학과                                                         
건축과        3.206349  4.301587  3.777778  3.603175  3.761905
건축설비소방과    3.152542  5.644068  4.016949  4.813559  4.389831
경영과        2.850746  4.880597  4.223881  4.432836  3.880597
기계과        3.109091  4.745455  3.836364  3.490909  2.727273
디지털전자과     2.857143  4.196429  3.875000  4.089286  4.071429
메카트로닉스과    2.607143  5.071429  3.857143  4.571429  4.107143
모바일인터넷과    3.042553  5.361702  4.361702  4.595745  4.085106
방송음향영상과    3.872727  5.327273  4.127273  4.200000  4.127273
비서사무행정과    3.226415  5.773585  4.792453  4.754717  4.452830
사회복지과      3.384615  5.076923  3.769231  4.288462  3.730769
산업경영과      3.080000  5.320000  4.720000  4.000000  4.120000
세무회계과      3.737705  5.032787  4.229508  4.459016  4.180328
스포츠지도과     3.056604  4.245283  3.660377  3.283019  3.396226
실내디자인과     3.416667  4.933333  4.383333  3.983333  4.400000
아동문헌정보과    3.725490  5.431373  4.019608 

In [168]:
score2_mean19.equals(score2_mean15)


True

df19 자료와 df15 자료의 2019 년도 부분 재학생 과목 평균점수도 같음을 확인.

---------------------
<span style="color:yellow">5. 기타</span>

In [153]:
df15[df15['학과']=='문헌정보과']

연도     학과 학생구분  응시자수       A과목       B과목       F과목       D과목       G과목  \
115  2017  문헌정보과  신입생    77  3.415584  5.064935  3.558442  3.948052  4.103896   
143  2017  문헌정보과  재학생    24  3.708333  5.291667  3.541667  4.875000  4.708333   
171  2016  문헌정보과  신입생    77  3.415584  5.064935  3.558442  3.948052  4.103896   
199  2016  문헌정보과  재학생    24  3.708333  5.291667  3.541667  4.875000  4.708333   
241  2015  문헌정보과  신입생    33  3.606061  5.454545  3.121212  3.424242  4.636364   
270  2015  문헌정보과  재학생    31  3.516129  4.870968  4.032258  4.064516  4.806452   

     C과목  E과목  
115  0.0  0.0  
143  0.0  0.0  
171  0.0  0.0  
199  0.0  0.0  
241  0.0  0.0  
270  0.0  0.0

In [154]:
df15[df15['학과']=='아동문헌정보과']

연도       학과 학생구분  응시자수       A과목       B과목       F과목       D과목  \
11   2019  아동문헌정보과  신입생    62  3.451613  4.983871  0.000000  4.064516   
36   2019  아동문헌정보과  재학생    51  3.725490  5.431373  0.000000  4.862745   
64   2018  아동문헌정보과  신입생    34  3.176471  4.823529  3.176471  3.147059   
92   2018  아동문헌정보과  재학생    30  3.866667  5.133333  3.633333  4.700000   
106  2017  아동문헌정보과  신입생    74  3.190000  5.240000  3.490000  4.010000   
107  2017  아동문헌정보과  재학생    43  3.000000  5.000000  4.000000  4.000000   

          G과목       C과목       E과목  
11   0.000000  3.693548  4.064516  
36   0.000000  4.019608  4.254902  
64   3.176471  0.000000  0.000000  
92   4.533333  0.000000  0.000000  
106  4.190000  0.000000  0.000000  
107  4.000000  0.000000  0.000000

2017년을 기점으로 문헌정보과 가 이후 없어지고 아동문헌정보과 가 새로 나타나는데 정확한 관계를 알 수 없다.  
아동문헌정보과 의 2017 재학생 (index 107) 평균점수가 정수로 나타나는데 이는 정상적으로는 거의 불가능한 일이며, 또한 index 92, 106 의 평균점수에도 부자연스러운 규칙성이 보이므로 자료 제공자의 데이터 전처리 과정에서 무엇인가 문제가 있었을 가능성이 높다.  
기타, 주어진 "3. 평가시험 결과 보고서-템플릿.pdf" 자료에 몇몇 학과의 학부 과정 통폐합 정보가 담겨 있다.